src/ai/embeddings.rs

```diff
@@ -88,33 +87,32 @@ pub async fn get_related_macro_docs(
         .fetch_all(pool)
         .await?;
 
-    let duplicate_titles =
-        get_duplicate_titles(docs.clone().into_iter().map(|doc| doc.title).collect());
-
-    docs = docs
-        .into_iter()
-        .map(|doc| RelatedDoc {
-            title: match (duplicate_titles.contains(&doc.title), &doc.title_parent) {
-                (true, Some(title_parent)) => format!("{} ({})", doc.title, title_parent),
-                _ => doc.title,
-            },
-            ..doc
-        })
-        .collect();
+    let duplicate_titles = get_duplicate_titles(&docs);
+
+    docs.iter_mut().for_each(|doc| {
+        if let (true, Some(title_parent)) =
+            (duplicate_titles.contains(&doc.title), &doc.title_parent)
+        {
+            doc.title = format!("{} ({})", doc.title, title_parent);
+        }
+    });
 
     Ok(docs)
 }
 
-fn get_duplicate_titles(titles: Vec<String>) -> Vec<String> {
-    let mut counts: HashMap<String, u8> = HashMap::new();
-    for title in titles.iter() {
-        let count = counts.entry(title.to_string()).or_insert(0);
-        *count += 1;
-    }
-    counts
-        .into_iter()
-        .filter(|(_, count)| count > &1)
-        .map(|(title, _)| title)
+fn get_duplicate_titles<'a>(titles: &'a [RelatedDoc]) -> Vec<String> {
+    titles
+        .iter()
+        .map(|x| &x.title)
+        .sorted()
+        .dedup_by_with_count(|a, b| a == b)
```

Comment:
Where does `dedup_by_with_count()` come from? I don't see it in the stdlib.

Reply:
^ @fiji-flo

Reply:
Ah, found it: https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.dedup_by_with_count

Reply:
But looks like we can just use [duplicates()](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.duplicates) instead.